internal/cloud/gcp/cloudbuild.go

```diff
@@ -0,0 +1,224 @@
+	}
+	defer storageClient.Close()
+
+	getBuldReq := &cloudbuildpb.GetBuildRequest{
+		ProjectId: g.creds.ProjectID,
+		Id:        buildID,
+	}
+
+	imageBuild, err := cloudbuildClient.GetBuild(ctx, getBuldReq)
+	if err != nil {
+		return "", fmt.Errorf("failed to get the build info: %v", err)
+	}
+
+	// Determine the log file's Bucket and Object name
+	// Logs_bucket example: "gs://550072179371.cloudbuild-logs.googleusercontent.com"
+	// Logs file names will be of the format `${logs_bucket}/log-${build_id}.txt`
+	logBucket := imageBuild.LogsBucket
+	logBucket = strings.TrimPrefix(logBucket, "gs://")
+	// logBucket may can contain directory in its name if set to a custom value
+	var logObjectDir string
+	if strings.Contains(logBucket, "/") {
+		ss := strings.SplitN(logBucket, "/", 2)
+		logBucket = ss[0]
+		logObjectDir = fmt.Sprintf("%s/", ss[1])
+	}
+	logObject := fmt.Sprintf("%slog-%s.txt", logObjectDir, buildID)
+
+	// Read the log
+	logBuilder := new(strings.Builder)
+	rd, err := storageClient.Bucket(logBucket).Object(logObject).NewReader(ctx)
+	if err != nil {
+		return "", fmt.Errorf("failed to create a new Reader for object '%s/%s': %v", logBucket, logObject, err)
+	}
+	_, err = io.Copy(logBuilder, rd)
+	if err != nil {
+		return "", fmt.Errorf("reading data from object '%s/%s' failed: %v", logBucket, logObject, err)
+	}
+
+	return logBuilder.String(), nil
+}
+
+// CloudbuildBuildCleanup parses the logs for the specified Build job and tries to clean up all resources
+// which were created as part of the job. It returns list of strings with all resources that were deleted
+// as a result of calling this method.
+//
+// Uses:
+//	- Storage API
+//	- Cloud Build API
+//	- Compute Engine API (indirectly)
+func (g *GCP) CloudbuildBuildCleanup(ctx context.Context, buildID string) ([]string, error) {
+	var deletedResources []string
+
+	storageClient, err := storage.NewClient(ctx, option.WithCredentials(g.creds))
+	if err != nil {
+		return deletedResources, fmt.Errorf("failed to get Storage client: %v", err)
+	}
+	defer storageClient.Close()
+
+	buildLog, err := g.CloudbuildBuildLog(ctx, buildID)
+	if err != nil {
+		return deletedResources, fmt.Errorf("failed to get log for build ID '%s': %v", buildID, err)
+	}
+
+	resources, err := cloudbuildResourcesFromBuildLog(buildLog)
+	if err != nil {
+		return deletedResources, fmt.Errorf("extracting created resources from build log failed: %v", err)
+	}
+
+	// Delete all Compute Engine instances
+	for _, instance := range resources.computeInstances {
+		err = g.ComputeInstanceDelete(ctx, resources.zone, instance)
+		if err == nil {
+			deletedResources = append(deletedResources, fmt.Sprintf("instance: %s (%s)", instance, resources.zone))
+		}
+	}
+
+	// Deleting instances in reality takes some time. Deleting a disk while it is still used by the instance, will fail.
+	// Iterate over the list of instances and wait until they are all deleted.
+	for _, instance := range resources.computeInstances {
+		for {
```

Comment:
Did you consider a timeout for this loop?

Reply:
Not really. Instances are deleted in the step before and if the code does not wait for the GCP service to finish deleting them, then the subsequent deletion of disks will fail if the instance is still not deleted. While an instance can have "Deletion Protection" set and in this case we could end up in an infinite loop, it is not set on the instances created by the image import job.

I'm open to suggestion how to make the code more robust. However I don't think that timeout makes sense here. Mainly because timing out will potentially cause the subsequent disk deletions to fail. Maybe at least check whether the instance does not have deletion protection set?

Reply:
I added check if the instance is protected against deletion so prevent a potential infinite loop. I don't think that having a timeout here makes sense, but I'm open to more feedback and also to changing my opinion.

Reply:
Just to have a complete information, the GCP has a [shutdown period](https://cloud.google.com/compute/docs/instances/deleting-instance#delete_timeout), which is in most cases 90 seconds. If the instance does not shut down gracefully within that time, it will shut it down forcefully. Since all instances are deleted first and only after that the code waits for the shutdown to complete, the combined maximum time that it can take should be also around 90 seconds. The reason is that all of the instances will be effectively shutting down in parallel.

Reply:
Thanks for the detailed explanation!